python/src/array.cpp

```diff
@@ -770,37 +770,90 @@ void init_array(nb::module_& m) {
+              }
+            }
             return equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__lt__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return less(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__le__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return less_equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__gt__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return greater(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__ge__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return false in case of object comparison which is not mlx
+              // array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return false;
+              }
+            }
             return greater_equal(a, to_array(v, a.dtype()));
           },
           "other"_a)
       .def(
           "__ne__",
-          [](const array& a, const ScalarOrArray v) {
+          [](const array& a,
+             const ScalarOrArray v) -> std::variant<array, bool> {
+            auto check_for_obj = std::get_if<nb::object>(&v);
+            if (check_for_obj) {
+              // return true in case of object inequlity which is not mlx array
+              if (!isMlxCoreArray(*check_for_obj)) {
+                return true;
+              }
+            }
```

Comment:
This seems broken because it will no longer allow `==` between arrays and integers or floats.

Reply:
No. The equality between an array and a scalar still works. There's a test that I wrote and covers this aspect:
```
    def test_array_eq_scalar(self):
        a = mx.array([1, 2, 3])
        b = 1
        c = 4

        self.assertTrue(mx.any(a == b))
        self.assertFalse(mx.all(a == c))
```
These lines of code, check for if the argument is of type `object`, and if so, checks if it's of type `mlx.core.array`. If not, so it's a list, tuple or an instance of a class, compared to an array.

Reply:
I mean it doesn't really make sense but I do see that you are right. I would have thought scalars were considered `nb::object` as well, that is surprising..

Reply:
Evidently its not so difficult to check if something is convertible to an `array` so I might actually do the check explicitly rather than use the variant return type. That may be a premature optimization.

But I would prefer you have a standalone function in `utils.h` like this:

```
bool is_convertible_to_array(const ScalarOrArray& v) {
 // Do the check here and return true/false
}
```

Reply:
Done ✅